Documentation: add vmTools

kip93 opened the pull request:
###### Description of changes

#190809 states that there was no docs for `runInLinuxVM`, but when looking at the manual I found that the entire `vmTools` set is missing. I tried extracting the comments from [that file](https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/vm/default.nix) into the docs, do some reformatting, and finally test and add some examples of my own.

Only thing missing are `extractFs` and `extractMTDfs`, for which I can't think of any clear / concise examples.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

kip93 committed:
doc: add vmTools

kip93 committed:
Remove unused reference

fricklerhandwerk commented on doc/builders/special/vm-tools.section.md:
````diff
@@ -0,0 +1,116 @@
+
+### Examples
+
+```nix
+# Build the derivation hello inside a VM
+runInLinuxVM hello
+
+# Build inside a VM with extra memory
+runInLinuxVM (hello.overrideAttrs (_: { memSize = 1024; }))
+
+# Use VM with a disk image (implicitly sets `diskImage`, see `vmTools.createEmptyImage`)
+runInLinuxVM (hello.overrideAttrs (_: { preVM = createEmptyImage { size = 1024; fullName = "vm-image"; }; }))
+```
````
Do not use prose in comments, it's hard to read and to discern from *explanatory* comments as we're used to in actual code. 

Not sure about the scope aspect. While the samples are nice and easy on the eye, they're not self-contained and therefore not usable without contextual knowledge that we cannot assume from beginners (even if you say this is for advanced use cases - no one knows: that may be the first thing someone does with Nix, for whatever reason). 

If we had Sphinx, we could have full working samples and highlight the relevant part. I leave it up to you to decorate every one with `{ pkgs }: with pkgs;` or not.

````suggestion
Build the derivation hello inside a VM:

```nix
runInLinuxVM hello
```

Build inside a VM with extra memory:

```nix
runInLinuxVM (hello.overrideAttrs (_: { memSize = 1024; }))
```

Use VM with a disk image (implicitly sets `diskImage`, see 
[`vmTools.createEmptyImage`](#vm-tools-createEmptyImage)):

```nix
runInLinuxVM (hello.overrideAttrs (_: {
  preVM = createEmptyImage {
    size = 1024;
    fullName = "vm-image";
  };
}))
```
````

fricklerhandwerk commented on doc/builders/special/vm-tools.section.md:
````diff
@@ -0,0 +1,116 @@
+runInLinuxVM (hello.overrideAttrs (_: { preVM = createEmptyImage { size = 1024; fullName = "vm-image"; }; }))
+```
+
+<!-- TODO
````
(nit) We don't have a TODO/FIXME policy, but I have learned to keep FIXME's for posterity and TODO's for oneself. Therefore: finish or leave for posterity.

```suggestion
<!-- FIXME
```

fricklerhandwerk commented on doc/builders/special/vm-tools.section.md:
```diff
@@ -0,0 +1,116 @@
+
+## `vmTools.runInLinuxImage` {#vm-tools-runInLinuxImage}
+
+Like [](#vm-tools-runInLinuxVM), but run the build not using the `stdenv` from the Nix store, but using the tools
```
```suggestion
Like [](#vm-tools-runInLinuxVM), but instead of using `stdenv` from the Nix store, run the build using the tools
```

fricklerhandwerk commented on doc/builders/special/vm-tools.section.md:
```diff
@@ -0,0 +1,116 @@
+
+Like [](#vm-tools-runInLinuxVM), but run the build not using the `stdenv` from the Nix store, but using the tools
+provided by `/bin`, `/usr/bin`, etc. from the specified filesystem image, which typically is a filesystem containing a
+non-NixOS Linux distribution.
```
```suggestion
[FHS](https://en.wikipedia.org/wiki/Filesystem_Hierarchy_Standard)-based Linux distribution.
```

[I think we should avoid the term *non-NixOS*.](https://discourse.nixos.org/t/should-we-give-a-name-to-nix-on-non-nixos/21020/3)

fricklerhandwerk commented on doc/builders/special/vm-tools.section.md:
````diff
@@ -0,0 +1,116 @@
+
+### Examples
+
+```nix
+# Create a script for running a Fedora 27 VM
+makeImageTestScript diskImages.fedora27x86_64
+
+# Create a script for running an Ubuntu 20.04 VM
+makeImageTestScript diskImages.ubuntu2004x86_64
+```
````
As above.

````suggestion
Create a script for running a Fedora 27 VM:

```nix
makeImageTestScript diskImages.fedora27x86_64
```

Create a script for running an Ubuntu 20.04 VM:

```nix
makeImageTestScript diskImages.ubuntu2004x86_64
```
````

fricklerhandwerk commented on doc/builders/special/vm-tools.section.md:
```diff
@@ -0,0 +1,116 @@
+
+### Attributes
+
+* `size` (optional, defaults to `4096`). The size of the image, in megabytes.
```
As above.

```suggestion
* `size` (optional, defaults to `4096`). The size of the image, in MiB.
```

fricklerhandwerk commented on doc/builders/special/vm-tools.section.md:
```diff
@@ -0,0 +1,116 @@
+### Attributes
+
+* `size` (optional, defaults to `4096`). The size of the image, in megabytes.
+* `extraPackages` (optional). A list names of additional packages from the distribution that should be included in the
+  image.
```
Please don't wrap lines arbitrarily. I know, this is a contentious topic. The documentation team has agreed to roll out a [convention to break lines at sentence boundaries](https://github.com/NixOS/nix.dev/blob/master/CONTRIBUTING.md?plain=1#L175-L185) for the simple reason that it's easier to edit and diff.

```suggestion
* `extraPackages` (optional). A list names of additional packages from the distribution that should be included in the image.
```

Not super important here, and I won't ask you to reformat everything, just to make you aware.

fricklerhandwerk commented on doc/builders/special/vm-tools.section.md:
````diff
@@ -0,0 +1,116 @@
+
+### Examples
+
+```nix
+# 8G image containing Firefox in addition to the default packages.
+diskImageFuns.ubuntu2004x86_64 { extraPackages = [ "firefox" ]; size = 8192; }
+```
````
````suggestion
8GiB image containing Firefox in addition to the default packages:

```nix
diskImageFuns.ubuntu2004x86_64 { extraPackages = [ "firefox" ]; size = 8192; }
```
````

fricklerhandwerk reviewed (approved):
@kip93 thanks for this great contribution, really well done. Sorry that you had to wait so much, I only now got to work off my backlog.

I left a few change requests concerning formatting, but in terms of content and wording this is exemplary. I like it a lot. We need more of that kind.

The structure of the attribute listings strikes me as a bit unusual, but we haven't converged on a consistent style across the manual anyway, so it's not really important.

kip93 committed:
Update vmTools docs after rebase

kip93 commented:
Sorry for taking so long, this PR slipped my mind for the longest time.

I've rebased and updated to keep up to date with the changes in the source code, plus applied the suggested changes. The only standing issue is with the examples. Don't know how to improve them to make them more friendly to a new user, these seem to me clear for the topic at hand, and cover the more common scenarios that I can think of.

fricklerhandwerk reviewed (approved):
Looks awesome, thanks a lot for keeping at it!
